### ISSUE TITLE
[InstCombine] Wrong signed zero optimization

### ISSUE THREAD

**Issue by zhroma** [llvm/llvm-project#57357]
The following C snippet (see also https://godbolt.org/z/E41zbrThP)
```c
double f(double x) {
    return x < 0.0 ? -0.0 : x;
}
int main(void) {
    printf("%g\n", 1.0/f(0.0));
}
```
prints `-inf` when compiled with `-O1` or `-O2`. Right answer is positive `inf`.

Clang versions since 8.0.0 and newer are affected, 7.1.0 works fine.

The problem is inside `InstCombine` which transforms 
```llvm
define dso_local double @f(double %0) local_unnamed_addr #0 {
  %2 = fcmp olt double %0, 0.000000e+00
  %3 = select i1 %2, double -0.000000e+00, double %0
  ret double %3
}
```
into
```llvm
define dso_local double @f(double %0) local_unnamed_addr #0 {
  %2 = fcmp ole double %0, 0.000000e+00
  %3 = select i1 %2, double -0.000000e+00, double %0
  ret double %3
}
```
This is incorrect https://alive2.llvm.org/ce/z/jFq6m7

```console
Transformation doesn't verify!

ERROR: Value mismatch

Example:
double %0 = #x0000000000000000 (+0.0)

Source:
i1 %2 = #x0 (0)
double %3 = #x0000000000000000 (+0.0)

Target:
i1 %2 = #x1 (1)
double %3 = #x8000000000000000 (-0.0)
Source value: #x0000000000000000 (+0.0)
Target value: #x8000000000000000 (-0.0)
```

It would be correct only when `select` has `nsz` fast-math flag.

#49625 tells probably about the same bug, but example here is much simpler.

**Comment by rotateright** [comment]
The problem transform is:
IC: Visiting:   %2 = fcmp olt double %0, 0.000000e+00
IC: Visiting:   %3 = select i1 %2, double -0.000000e+00, double %0
ADD DEFERRED:   %3 = fcmp ugt double %0, -0.000000e+00
ADD DEFERRED:   %4 = select i1 %3, double %0, double -0.000000e+00

This is caused by ValueTracking identifying the code here as a MAXNUM pattern. There's already a FIXME note in there, so that's probably where we should solve it.

**Comment by rotateright** [comment]
Thanks for filing the bug report. This example should be fixed with 4e44c22c97be9. If you find any other cases like this, please open another issue.

**Comment by zhroma** [comment]
Many thanks for fixing this so fast!
